consensus/src/txn_hash_and_authenticator_deduper.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{TXN_DEDUP_FILTERED, TXN_DEDUP_SECONDS},
+    transaction_deduper::TransactionDeduper,
+};
+use aptos_types::transaction::SignedTransaction;
+use rayon::prelude::*;
+use std::collections::{HashMap, HashSet};
+
+pub struct TxnHashAndAuthenticatorDeduper {}
```

Comment:
(1) Why does this need to be public? As long as the TransactionDeduperType is public, and since we use dyn traits, do we still need to have this (and the NoOpDeduper) public?

(2) Let's add a comment explaining the rationale for using both the hash and the signature (without signature we can have a txn with a wrong signature dedup out a correct one and then fail, and without hash since we check signatures later, we could have a similar problem - or do I remember correctly? also @sasha8).

(Unrelated to critical path for this PR) we could add a TODO at relevant places and file a corresponding task to move signature verification earlier from AptosVM / block executor wrapper (anyway useless for state-sync as we discussed w. @JoshLind) and then could use just signature (unless there was some other problem I don't recall). Can also revisit caching the hash when signature is verified. Not to add any work here, I could file that task if it makes sense to everybody.

Reply:
For (1) because the two are in different files, it has to be at least pub(crate)

Reply:
sorry for confusion, I also meant pub(crate) vs pub. Makes sense why not private.

Assuming the VM prologue still checked that the signature was from the right account, but we had previously checked that the signature is properly signed (with whoever's credentials) on the raw txn, then for de-dup purposes just signature should be Ok since that would also fix the raw txn (due to the proper signing check), right? (or am I still falling into a trap?)

Reply:
Hmm, so given this sequence:
1. Check signature is properly signed (with whoever's credentials)
2. Dedup based on only signature
3. VM prologue checks that signature is from the right account

Step 2 could have false positive dedups, allowing this attack: a bad client just replays a correct raw txn, but signing with it's own signature. This causes the good txn to be deduped in Step 2. The bad client's txn would be discarded in Step 3. So no version of the raw txn is executed; it has been censored.

Reply:
Thanks - that's it - different credentials! very tired today :) but yeah, makes sense to document for future memoryless braindead folks like me xD

Reply:
i thought i recalled this, but.. if credentials are different, wouldn't the signature be different in step 2?

Reply:
I think you’re right, we can use only the signatures once the signatures are verified 👍  Once verified, it’s guaranteed that for pairs (txn_0, sig_0), (txn_1, sig_1): if sig_0 == sig_1 then txn_0 == txn_1